src/audio_decoder.cpp

```diff
@@ -170,38 +181,32 @@ void AudioDecoder::Update(std::chrono::microseconds delta) {
 	}
 
 	fade_time -= delta;
-	volume += std::chrono::duration_cast<std::chrono::milliseconds>(delta).count() * delta_volume_step;
 
-	volume = volume > 100.0 ? 100.0 :
-		volume < 0.0 ? 0.0 :
-		volume;
+	volume += static_cast<double>(std::chrono::duration_cast<std::chrono::microseconds>(delta).count()) * delta_volume_step;
+	volume = Utils::Clamp<double>(static_cast<double>(volume), 0.0, 100.0);
+	log_volume = AdjustVolume(volume);
 }
 
 int AudioDecoder::GetVolume() const {
-	return (int)volume;
+	return static_cast<int>(log_volume);
 }
 
 void AudioDecoder::SetVolume(int volume) {
-	this->volume = (double)volume;
+	this->volume = Utils::Clamp<double>(static_cast<double>(volume), 0.0, 100.0);
+	this->log_volume = AdjustVolume(this->volume);
 }
```

Comment:
I think this would read easier without `this->`.

```suggestion
void AudioDecoder::SetVolume(int new_volume) {
	volume = Utils::Clamp<double>(static_cast<double>(new_volume), 0.0, 100.0);
	log_volume = AdjustVolume(volume);
}
```

Sugestion feature merges somehow old and new code 🤦🏼